plan/statistics_helper.go

```diff
@@ -0,0 +1,166 @@
+	totalCount := int64(0)
+	for _, indexRange := range indexRanges {
+		lv := indexRange.LowVal
+		rv := indexRange.HighVal
+		for i := len(lv); i < len(indexInfo.Columns); i++ {
+			lv = append(lv, types.MinNotNullDatum())
+		}
+		for i := len(rv); i < len(indexInfo.Columns); i++ {
+			rv = append(rv, types.MaxValueDatum())
+		}
+		lb, err := codec.EncodeKey(nil, lv...)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		rb, err := codec.EncodeKey(nil, rv...)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		l := types.NewBytesDatum(lb)
+		r := types.NewBytesDatum(rb)
+		rowCount, err := getRowCountByRange(sc, statsTbl.Count, statsTbl.Indices[offset], l, r)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		totalCount += rowCount
+	}
+	if totalCount > statsTbl.Count {
+		totalCount = statsTbl.Count
+	}
+	return uint64(totalCount), nil
+}
+
+func getPseudoRowCountByIndexRanges(sc *variable.StatementContext, statsTbl *statistics.Table, indexRanges []*IndexRange, indexInfo *model.IndexInfo) (uint64, error) {
+	totalCount := float64(0)
+	for _, indexRange := range indexRanges {
+		count := float64(statsTbl.Count)
+		i := len(indexRange.LowVal) - 1
+		l := indexRange.LowVal[i]
+		r := indexRange.HighVal[i]
+		offset := indexInfo.Columns[i].Offset
+		rowCount, err := getRowCountByRange(sc, statsTbl.Count, statsTbl.Columns[offset], l, r)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		count = count / float64(statsTbl.Count) * float64(rowCount)
+		// If the condition is a = 1, b = 1, c = 1, d = 1, we think every a=1, b=1, c=1 only filtrate 1/100 data,
+		// so as to avoid collapsing too fast.
+		for j := 0; j < i; j++ {
+			count = count / float64(100)
+		}
+		totalCount += count
+	}
+	// To avoid the totalCount become too small.
+	if uint64(totalCount) < 1000 {
+		// We will not let the row count less than 1000 to avoid collapsing too fast in the future calculation.
+		totalCount = 1000.0
+	}
+	if totalCount > float64(statsTbl.Count) {
+		totalCount = float64(statsTbl.Count) / 3.0
+	}
+	return uint64(totalCount), nil
+}
+
+func getRowCountByTableRange(sc *variable.StatementContext, statsTbl *statistics.Table, ranges []TableRange, offset int) (uint64, error) {
+	var rowCount uint64
+	for _, rg := range ranges {
+		var cnt int64
+		var err error
+		if rg.LowVal == math.MinInt64 && rg.HighVal == math.MaxInt64 {
+			cnt = statsTbl.Count
+		} else if rg.LowVal == math.MinInt64 {
+			cnt, err = statsTbl.Columns[offset].LessRowCount(sc, types.NewDatum(rg.HighVal))
+		} else if rg.HighVal == math.MaxInt64 {
+			cnt, err = statsTbl.Columns[offset].GreaterRowCount(sc, types.NewDatum(rg.LowVal))
+		} else {
+			if rg.LowVal == rg.HighVal {
+				cnt, err = statsTbl.Columns[offset].EqualRowCount(sc, types.NewDatum(rg.LowVal))
+			} else {
+				cnt, err = statsTbl.Columns[offset].BetweenRowCount(sc, types.NewDatum(rg.LowVal), types.NewDatum(rg.HighVal))
+			}
+		}
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
```

Comment:
If the calculated count is greater than (HighVal - LowVal), we can set it to (HighVal - LowVal).
Should consider (MaxInt64 - MinInt64) overflow int64.